runtime/gc_vlhgc/CopyForwardCompactGroup.hpp

```diff
@@ -126,12 +132,141 @@ class MM_CopyForwardCompactGroup
 	{
 		_TLHRemainderBase = NULL;
 		_TLHRemainderTop = NULL;
+		_gcIDForTLHRemainder = 0;
 	}
 
 	MMINLINE void setTLHRemainder(void *base, void *top)
 	{
 		_TLHRemainderBase = base;
 		_TLHRemainderTop = top;
+		_TLHRemainderCount += 1;
+	}
+
+	MMINLINE void moveTLHRemainder(MM_CopyForwardCompactGroup *dest)
+	{
+		dest->_TLHRemainderBase = _TLHRemainderBase;
+		dest->_TLHRemainderTop = _TLHRemainderTop;
+		dest->_gcIDForTLHRemainder = _gcIDForTLHRemainder;
+		resetTLHRemainder();
+	}
+
+	MMINLINE uintptr_t getTLHRemainderSize()
+	{
+		return ((uintptr_t)_TLHRemainderTop - (uintptr_t)_TLHRemainderBase);
+	}
+
+	/**
+	 * return true if TLHRemainder has been aligned.
+	 */
+	MMINLINE bool alignTLHRemainder(uintptr_t tlhSurvivorDiscardThreshold)
+	{
+		bool aligned = false;
+		if (_TLHRemainderBase) {
+			void *newAlignedTLHRemainderBase = (void*) MM_Math::roundToCeilingCard((uintptr_t) _TLHRemainderBase);
+			void *newAlignedTLHRemainderTop =  (void*) MM_Math::roundToFloorCard((uintptr_t) _TLHRemainderTop);
+			if (((uintptr_t)newAlignedTLHRemainderTop - (uintptr_t)newAlignedTLHRemainderBase) >= tlhSurvivorDiscardThreshold) {
+				/* TLHRemainder need to be card aligned for being reuse cross PGCs */
+				_TLHRemainderBase = newAlignedTLHRemainderBase;
+				_TLHRemainderTop = newAlignedTLHRemainderTop;
+				aligned = true;
+			}
+			/* else aligned TLHRemainder is smaller than tlhSurvivorDiscardThreshold, need to be reset */
+		}
+		return aligned;
+	}
+
+	MMINLINE void ageTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t age, MM_CopyForwardCompactGroup *dest)
+	{
```

Comment:
I think you don't have to pass dest, since you easily find it here as _this + 1_. And I'd call it destGroupCandidate:

MM_CopyForwardCompactGroup *destGroupCandidate = this + 1;